File: scala/src/main/scala/xieyuheng/cicada/grammar.scala
package xieyuheng.cicada

import collection.immutable.ListMap

import xieyuheng.partech._
import xieyuheng.partech.ruleDSL._
import xieyuheng.partech.predefined._

object grammar {

  val lexer = common_lexer

  val preserved =
    List(
      "type", "class",
      "case",
      "string_t",
    )

  val identifier = identifier_with_preserved("identifier", preserved)

  val top_entry: () => Rule =
    () => Rule(
      "top_entry", Map(
        "block_entry" -> List(block_entry),
        "@refuse" -> List("@", "refuse", exp, ":", exp),
        "@accept" -> List("@", "accept", exp, ":", exp),
        "@show" -> List("@", "show", exp),
        "@eq" -> List("@", "eq", exp, "=", exp),
      ))

  def top_entry_matcher: Tree => Top =
    Tree.matcher[Top](
      "top_entry", Map(
        "block_entry" -> { case List(block_entry) =>
          block_entry_matcher(block_entry) match {
            case (name, block_entry) =>
              block_entry match {
                case BlockEntryLet(value: Exp) =>
                  TopLet(name, value)
                case BlockEntryDefine(t: Exp, value: Exp) =>
                  TopDefine(name, t, value)
              }
          } },
        "@refuse" -> { case List(_, _, exp, _, t) =>
          TopKeywordRefuse(exp_matcher(exp), exp_matcher(t)) },
        "@accept" -> { case List(_, _, exp, _, t) =>
          TopKeywordAccept(exp_matcher(exp), exp_matcher(t)) },
        "@show" -> { case List(_, _, exp) =>
          TopKeywordShow(exp_matcher(exp)) },
        "@eq" -> { case List(_, _, rhs, _, lhs) =>
          TopKeywordEq(exp_matcher(rhs), exp_matcher(lhs)) },
      ))

  val top_list: Rule = non_empty_list(List(top_entry))

  def top_list_matcher = non_empty_list_matcher(top_entry_matcher)

  val exp: () => Rule =
    () => Rule(
      "exp", Map(
        "var" -> List(identifier),
        "type" -> List("type"),
        "string_t" -> List("string_t"),
        "string" -> List(double_quoted_string),
        "pi" -> List("{", $(non_empty_list, given_entry), "-", ">", exp, "}"),
        "fn" -> List("{", $(non_empty_list, given_entry), "=", ">", exp, "}"),
        "fn_case" -> List("{", $(non_empty_list, fn_case_clause), "}"),
        "ap" -> List(exp, "(", $(non_empty_list, arg_entry), ")"),
        "cl" -> List("class", "{", $(non_empty_list, given_entry), "}"),
        "cl_predefined" -> List("class", "{",
          $(non_empty_list, define_entry),
          $(non_empty_list, given_entry),
          "}"),
        "cl_predefined_empty_given" -> List("class", "{", $(non_empty_list, define_entry), "}"),
        "cl_naked" -> List("{", $(non_empty_list, given_entry), "}"),
        "cl_empty" -> List("class", "{", "}"),
        "obj" -> List("object", "{", $(non_empty_list, let_entry), "}"),
        "obj_naked" -> List("{", $(non_empty_list, let_entry), "}"),
        "obj_empty" -> List("object", "{", "}"),
        "obj_naked_empty" -> List("{", "}"),
        "dot" -> List(exp, ".", identifier),
        "block" -> List("{", $(non_empty_list, block_entry), exp, "}"),
        "the" -> List("the", "(", exp, ",", exp, ")"),
      ))

  def exp_matcher: Tree => Exp =
    Tree.matcher[Exp](
      "exp", Map(
        "var" -> { case List(Leaf(name)) => Var(name.word) },
        "type" -> { case List(_) => Type() },
        "string_t" -> { case List(_) => StrType() },
        "string" -> { case List(Leaf(str)) =>
          Str(trim_double_quote(str.word)) },
        "pi" -> { case List(_, given_entry_list, _, _, return_type, _) =>
          val type_map = ListMap(non_empty_list_matcher(given_entry_matcher)(given_entry_list): _*)
          Pi(type_map, exp_matcher(return_type)) },
        "fn" -> { case List(_, given_entry_list, _, _, body, _) =>
          val type_map = ListMap(non_empty_list_matcher(given_entry_matcher)(given_entry_list): _*)
          Fn(type_map, exp_matcher(body)) },
        "fn_case" -> { case List(_, fn_case_clause_list, _) =>
          FnCase(non_empty_list_matcher(fn_case_clause_matcher)(fn_case_clause_list)) },
        "ap" -> { case List(target, _, arg_entry_list, _) =>
          val args = non_empty_list_matcher(arg_entry_matcher)(arg_entry_list)
          Ap(exp_matcher(target), args) },
        "cl" -> { case List(_, _, given_entry_list, _) =>
          val type_map = ListMap(non_empty_list_matcher(given_entry_matcher)(given_entry_list): _*)
          Cl(ListMap.empty, type_map) },
        "cl_predefined" -> { case List(
          _, _, define_entry_list, given_entry_list, _) =>
          val defined = ListMap(non_empty_list_matcher(define_entry_matcher)(define_entry_list): _*)
          val type_map = ListMap(non_empty_list_matcher(given_entry_matcher)(given_entry_list): _*)
          Cl(defined, type_map) },
        "cl_predefined_empty_given" -> { case List(
          _, _, define_entry_list, _) =>
          val defined = ListMap(non_empty_list_matcher(define_entry_matcher)(define_entry_list): _*)
          Cl(defined, ListMap()) },
        "cl_naked" -> { case List(_, given_entry_list, _) =>
          val type_map = ListMap(non_empty_list_matcher(given_entry_matcher)(given_entry_list): _*)
          Cl(ListMap.empty, type_map) },
        "cl_empty" -> { case List(_, _, _) =>
          Cl(ListMap.empty, ListMap.empty) },
        "obj" -> { case List(_, _, let_entry_list, _) =>
          val value_map = ListMap(non_empty_list_matcher(let_entry_matcher)(let_entry_list): _*)
          Obj(value_map) },
        "obj_naked" -> { case List(_, let_entry_list, _) =>
          val value_map = ListMap(non_empty_list_matcher(let_entry_matcher)(let_entry_list): _*)
          Obj(value_map) },
        "obj_empty" -> { case List(_, _, _) =>
          Obj(ListMap.empty) },
        "obj_naked_empty" -> { case List(_, _) =>
          Obj(ListMap()) },
        "dot" -> { case List(target, _, Leaf(field_name)) =>
          Dot(exp_matcher(target), field_name.word) },
        "block" -> { case List(_, block_entry_list, body, _) =>
          val block_entry_map = ListMap(non_empty_list_matcher(block_entry_matcher)(block_entry_list): _*)
          Block(block_entry_map, exp_matcher(body)) },
      ))

  val fn_case_clause: () => Rule =
    () => Rule(
      "fn_case_clause", Map(
        "case" -> List("case", $(non_empty_list, given_entry), "=", ">", exp),
      ))

  def fn_case_clause_matcher =
    Tree.matcher[(ListMap[String, Exp], Exp)](
      "fn_case_clause", Map(
        "case" -> { case List(_, given_entry_list, _, _, body) =>
          val type_map = ListMap(non_empty_list_matcher(given_entry_matcher)(given_entry_list): _*)
          (type_map, exp_matcher(body)) },
      ))

  val arg_entry: () => Rule =
    () => Rule(
      "arg_entry", Map(
        "arg" -> List(exp),
        "arg_comma" -> List(exp, ","),
      ))

  def arg_entry_matcher =
    Tree.matcher[Exp](
      "arg_entry", Map(
        "arg" -> { case List(exp) => exp_matcher(exp) },
        "arg_comma" -> { case List(exp, _) => exp_matcher(exp) },
      ))

  val given_entry: () => Rule =
    () => Rule(
      "given_entry", Map(
        "given" -> List(identifier, ":", exp),
        "given_comma" -> List(identifier, ":", exp, ","),
      ))

  def given_entry_matcher =
    Tree.matcher[(String, Exp)](
      "given_entry", Map(
        "given" -> { case List(Leaf(name), _, exp) => (name.word, exp_matcher(exp)) },
        "given_comma" -> { case List(Leaf(name), _, exp, _) => (name.word, exp_matcher(exp)) },
      ))

  val let_entry: () => Rule =
    () => Rule(
      "let_entry", Map(
        "let" -> List(identifier, "=", exp),
        "let_comma" -> List(identifier, "=", exp, ","),
      ))

  def let_entry_matcher =
    Tree.matcher[(String, Exp)](
      "let_entry", Map(
        "let" -> { case List(Leaf(name), _, exp) => (name.word, exp_matcher(exp)) },
        "let_comma" -> { case List(Leaf(name), _, exp, _) => (name.word, exp_matcher(exp)) },
      ))

  val define_entry: () => Rule =
    () => Rule(
      "define_entry", Map(
        "define" -> List(identifier, ":", exp, "=", exp),
      ))

  def define_entry_matcher =
    Tree.matcher[(String, (Exp, Exp))](
      "define_entry", Map(
        "define" -> { case List(Leaf(name), _, t, _, exp) =>
          (name.word, (exp_matcher(t), exp_matcher(exp))) },
      ))

  val block_entry: () => Rule =
    () => Rule(
      "block_entry", Map(
        "let" -> List(identifier, "=", exp),
        "let_cl" -> List("class", identifier, "{",
          $(non_empty_list, given_entry),
          "}"),
        "let_cl_predefined" -> List("class", identifier, "{",
          $(non_empty_list, define_entry),
          $(non_empty_list, given_entry),
          "}"),
        "let_cl_predefined_empty_given" -> List("class", identifier, "{",
          $(non_empty_list, define_entry),
          "}"),
        "let_cl_empty" -> List("class", identifier, "{", "}"),
        "let_obj" -> List("object", identifier, "{", $(non_empty_list, let_entry), "}"),
        "let_obj_empty" -> List("object", identifier, "{", "}"),
        "define" -> List(identifier, ":", exp, "=", exp),
      ))

  def block_entry_matcher: Tree => (String, BlockEntry) =
    Tree.matcher[(String, BlockEntry)](
      "block_entry", Map(
        "let" -> { case List(Leaf(name), _, exp) =>
          (name.word, BlockEntryLet(exp_matcher(exp))) },
        "let_cl" -> { case List(_, Leaf(name), _, given_entry_list, _) =>
          val type_map = ListMap(non_empty_list_matcher(given_entry_matcher)(given_entry_list): _*)
          (name.word, BlockEntryDefine(Type(), Cl(ListMap.empty, type_map))) },
        "let_cl_predefined" -> { case List(_, Leaf(name), _,
          define_entry_list,
          given_entry_list,
          _) =>
          val defined = ListMap(non_empty_list_matcher(define_entry_matcher)(define_entry_list): _*)
          val type_map = ListMap(non_empty_list_matcher(given_entry_matcher)(given_entry_list): _*)
          (name.word, BlockEntryDefine(Type(), Cl(defined, type_map))) },
        "let_cl_predefined_empty_given" -> { case List(_, Leaf(name), _,
          define_entry_list,
          _) =>
          val defined = ListMap(non_empty_list_matcher(define_entry_matcher)(define_entry_list): _*)
          (name.word, BlockEntryDefine(Type(), Cl(defined, ListMap.empty))) },
        "let_cl_empty" -> { case List(_, Leaf(name), _, _) =>
          (name.word, BlockEntryDefine(Type(), Cl(ListMap.empty, ListMap.empty))) },
        "let_obj" -> { case List(_, Leaf(name), _, let_entry_list, _) =>
          val value_map = ListMap(non_empty_list_matcher(let_entry_matcher)(let_entry_list): _*)
          (name.word, BlockEntryLet(Obj(value_map))) },
        "let_obj_empty" -> { case List(_, Leaf(name), _, _) =>
          (name.word, BlockEntryLet(Obj(ListMap.empty))) },
        "define" -> { case List(Leaf(name), _, t, _, exp) =>
          (name.word, BlockEntryDefine(exp_matcher(t), exp_matcher(exp))) },
      ))

}

File: lib/grammar.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ptc = __importStar(require("@forchange/partech"));
const Top = __importStar(require("./top"));
const Exp = __importStar(require("./exp"));
const Scope = __importStar(require("./scope"));
const preserved = [
    "type", "class",
    "case",
    "string_t",
    "equation_t", "same", "transport",
];
const identifier = ptc.identifier_with_preserved("identifier", preserved);
const arg_entry = new ptc.Rule("arg_entry", {
    "arg": [exp],
    "arg_comma": [exp, ","],
});
const arg_entry_matcher = ptc.Node.matcher_with_span(span => [
    "arg_entry", {
        "arg": ([exp]) => exp_matcher(exp),
        "arg_comma": ([exp, _]) => exp_matcher(exp),
    }
]);
const top_entry = new ptc.Rule("top_entry", {
    "named_entry": [named_entry],
    "@refuse": ["@", "refuse", exp, ":", exp],
    "@accept": ["@", "accept", exp, ":", exp],
    "@show": ["@", "show", exp],
    "@eq": ["@", "eq", exp, "=", exp],
});
const top_entry_matcher = ptc.Node.matcher_with_span(span => [
    "top_entry", {
        "named_entry": ([named_entry]) => new Top.TopNamedScopeEntry(...named_entry_matcher(named_entry)),
        "@refuse": ([, , exp, , t]) => new Top.TopKeywordRefuse(exp_matcher(exp), exp_matcher(t)),
        "@accept": ([, , exp, , t]) => new Top.TopKeywordAccept(exp_matcher(exp), exp_matcher(t)),
        "@show": ([, , exp]) => new Top.TopKeywordShow(exp_matcher(exp)),
        "@eq": ([, , lhs, , rhs]) => new Top.TopKeywordEq(exp_matcher(lhs), exp_matcher(rhs)),
    }
]);
exports.top_list = new ptc.Rule("top_list", {
    "top_entry_list": [ptc.ap(ptc.non_empty_list, top_entry)],
});
exports.top_list_matcher = ptc.Node.matcher_with_span(span => [
    "top_list", {
        "top_entry_list": ([top_entry_list]) => ptc.non_empty_list_matcher(top_entry_matcher)(top_entry_list),
    }
]);
const scope = new ptc.Rule("scope", {
    "named_entries": [ptc.ap(ptc.non_empty_list, named_entry)],
});
const scope_matcher = ptc.Node.matcher_with_span(span => [
    "scope", {
        "named_entries": ([named_entries]) => new Scope.Scope(ptc.non_empty_list_matcher(named_entry_matcher)(named_entries)),
    }
]);
function named_entry() {
    return new ptc.Rule("named_entry", {
        "let": [identifier, "=", exp],
        "let_comma": [identifier, "=", exp, ","],
        "given": [identifier, ":", exp],
        "given_comma": [identifier, ":", exp, ","],
        "define": [identifier, ":", exp, "=", exp],
        "define_comma": [identifier, ":", exp, "=", exp, ","],
        "cl": ["class", identifier, "{", scope, "}"],
        "cl_empty": ["class", identifier, "{", "}"],
    });
}
const named_entry_matcher = ptc.Node.matcher_with_span(span => [
    "named_entry", {
        "let": ([name, , exp]) => [ptc.Leaf.word(name), new Scope.Entry.Let(exp_matcher(exp))],
        "let_comma": ([name, , exp,]) => [ptc.Leaf.word(name), new Scope.Entry.Let(exp_matcher(exp))],
        "given": ([name, , t]) => [ptc.Leaf.word(name), new Scope.Entry.Given(exp_matcher(t))],
        "given_comma": ([name, , t,]) => [ptc.Leaf.word(name), new Scope.Entry.Given(exp_matcher(t))],
        "define": ([name, , t, , exp]) => [ptc.Leaf.word(name), new Scope.Entry.Define(exp_matcher(t), exp_matcher(exp))],
        "define_comma": ([name, , t, , exp,]) => [ptc.Leaf.word(name), new Scope.Entry.Define(exp_matcher(t), exp_matcher(exp))],
        "cl": ([, name, , scope,]) => [ptc.Leaf.word(name), new Scope.Entry.Define(new Exp.Type(), new Exp.Cl(scope_matcher(scope)))],
        "cl_empty": ([, name, ,]) => [ptc.Leaf.word(name), new Scope.Entry.Define(new Exp.Type(), new Exp.Cl(new Scope.Scope()))],
    }
]);
function exp() {
    return new ptc.Rule("exp", {
        "var": [identifier],
        "type": ["type"],
        "string_t": ["string_t"],
        "string": [ptc.double_quoted_string],
        "pi": ["{", scope, "-", ">", exp, "}"],
        "fn": ["{", scope, "=", ">", exp, "}"],
        "fn_case": ["{", ptc.ap(ptc.non_empty_list, fn_case_clause), "}"],
        "ap": [exp, "(", ptc.ap(ptc.non_empty_list, arg_entry), ")"],
        "cl": ["class", "{", scope, "}"],
        "cl_empty": ["class", "{", "}"],
        "obj": ["{", scope, "}"],
        "obj_empty": ["{", "}"],
        "dot": [exp, ".", identifier],
        "block": ["{", scope, exp, "}"],
        "equation": ["equation_t", "(", exp, ",", exp, ",", exp, ")"],
        "same": ["same", "(", exp, ",", exp, ")"],
        "transport": ["transport", "(", exp, ",", exp, ",", exp, ")"],
    });
}
const exp_matcher = ptc.Node.matcher_with_span(span => [
    "exp", {
        "var": ([name]) => new Exp.Var(ptc.Leaf.word(name)),
        "type": _ => new Exp.Type(),
        "string_t": _ => new Exp.StrType(),
        "string": ([str]) => {
            let s = ptc.Leaf.word(str);
            return new Exp.Str(s.slice(1, s.length - 1));
        },
        "pi": ([, scope, , , return_type, _]) => new Exp.Pi(scope_matcher(scope), exp_matcher(return_type)),
        "fn": ([, scope, , , body, _]) => new Exp.Fn(scope_matcher(scope), exp_matcher(body)),
        "fn_case": ([, fn_case_clause_list, _]) => new Exp.FnCase(ptc.non_empty_list_matcher(fn_case_clause_matcher)(fn_case_clause_list)),
        "ap": ([target, , arg_entry_list, _]) => new Exp.Ap(exp_matcher(target), ptc.non_empty_list_matcher(arg_entry_matcher)(arg_entry_list)),
        "cl": ([, , scope, _]) => new Exp.Cl(scope_matcher(scope)),
        "cl_empty": _ => new Exp.Cl(new Scope.Scope()),
        "obj": ([, scope, _]) => new Exp.Obj(scope_matcher(scope)),
        "obj_empty": _ => new Exp.Obj(new Scope.Scope()),
        "dot": ([target, , field_name]) => new Exp.Dot(exp_matcher(target), ptc.Leaf.word(field_name)),
        "block": ([, scope, body, _]) => new Exp.Block(scope_matcher(scope), exp_matcher(body)),
        "equation": ([, , t, , lhs, , rhs,]) => new Exp.Equation(exp_matcher(t), exp_matcher(lhs), exp_matcher(rhs)),
        "same": ([, , t, , value,]) => new Exp.Same(exp_matcher(t), exp_matcher(value)),
        "transport": ([, , equation, , motive, , base,]) => new Exp.Transport(exp_matcher(equation), exp_matcher(motive), exp_matcher(base)),
    }
]);
const fn_case_clause = new ptc.Rule("fn_case_clause", {
    "case": ["case", scope, "=", ">", exp],
});
const fn_case_clause_matcher = ptc.Node.matcher_with_span(span => [
    "fn_case_clause", {
        "case": ([, scope, , , body]) => new Exp.Fn(scope_matcher(scope), exp_matcher(body)),
    }
]);
